CompuCell3D/core/CompuCell3D/plugins/FocalPointPlasticity/FocalPointPlasticityLinkInventoryBase.h

```diff
@@ -0,0 +1,233 @@
+			}
+			if (cell1) {
+				FocalPointPlasticityTrackerData fppd = _link->getFPPTrackerData(cell1);
+				getFPPTrackerDataSet(cell1).insert(fppd);
+			}
+		}
+		// Remove a link without any additional internal work
+		void removeLinkNoChain(LinkType* _link) {
+			const FPPLinkID linkId = getLinkId(_link);
+			linkInventoryItr_t itr = linkInventory.find(linkId);
+			if (itr != linkInventory.end()) linkInventory.erase(linkId);
+		}
+		// Remove tracker data from cells
+		void removeTrackerData(LinkType* _link) {
+			FPPTrackerDataSet::iterator itr;
+			CellG* cell0 = const_cast<CellG*>(_link->getObj0());
+			CellG* cell1 = const_cast<CellG*>(_link->getObj1());
+			if (cell0) {
+				FPPTrackerDataSet &fppdSet = getFPPTrackerDataSet(cell0);
+				FocalPointPlasticityTrackerData fppd = _link->getFPPTrackerData(cell0);
+				itr = fppdSet.find(fppd);
+				if (itr != fppdSet.end()) fppdSet.erase(fppd);
+			}
+			if (cell1) {
+				FPPTrackerDataSet &fppdSet = getFPPTrackerDataSet(cell1);
+				FocalPointPlasticityTrackerData fppd = _link->getFPPTrackerData(cell1);
+				itr = fppdSet.find(fppd);
+				if (itr != fppdSet.end()) fppdSet.erase(fppd);
+			}
+		}
+
+		FPPInventory_t* getCellLinkInventory(CellG* _cell) { 
+			return &cellLinkInventoryTrackerAccessor->get(_cell->extraAttribPtr)->linkInv;
+		}
+
+		CellG* getCellGFromConst(const CellG* _cell) {
+			if (!_cell) return (CellG*)(0);
+			else {
+				CellG* _cellNC = potts->getCellInventory().getCellByIds(_cell->id, _cell->clusterId);
+				return _cellNC;
+			}
+		}
+
+	public:
+		FPPLinkInventoryBase() {}
+		FPPLinkInventoryBase(BasicClassAccessor<FPPLinkInventoryTracker<LinkType> >* _cellLinkInventoryTrackerAccessor, Potts3D* _potts, BasicClassAccessor<FocalPointPlasticityTracker>* _focalPointPlasticityTrackerAccessor)
+		{
+			cellLinkInventoryTrackerAccessor = _cellLinkInventoryTrackerAccessor;
+			potts = _potts;
+			focalPointPlasticityTrackerAccessor = _focalPointPlasticityTrackerAccessor;
+		}
+		virtual ~FPPLinkInventoryBase() {}
+
+		BasicClassAccessor<FocalPointPlasticityTracker> * getFocalPointPlasticityTrackerAccessorPtr() { return focalPointPlasticityTrackerAccessor; }
+		BasicClassAccessor<FPPLinkInventoryTracker<LinkType> >* getFocalPointPlasticityCellLinkInventoryTrackerAccessorPtr() { return cellLinkInventoryTrackerAccessor; }
+
+		virtual FPPTrackerDataSet& getFPPTrackerDataSet(CellG* _cell) { return FPPTrackerDataSet(); }
+
+		virtual int getLinkInventorySize() { return linkInventory.size(); }
+		linkInventory_t& getContainer() { return linkInventory; }
+		virtual linkInventoryItr_t linkInventoryBegin() { return linkInventory.begin(); }
+		virtual linkInventoryItr_t linkInventoryEnd() { return linkInventory.end(); }
+		virtual void incrementIterator(linkInventoryItr_t& _itr) { ++_itr; }
+		virtual void decrementIterator(linkInventoryItr_t& _itr) { --_itr; }
+
+		// Add a link to the link inventory and update internals
+		void addToInventory(LinkType* _link) {
```

Comment:
I woudl imagine that this function woudl take a link , add it to inventory and this is it. Why are we calling those extra functions?

Reply:
addTrackerData manages the FPPTrackerData for each cell, and will be removed soon. The remaining three calls actually do the adding of a link to the container (addLinkNoChain). You'll notice that each cell has its own link inventory of the same link type (accessed through getCellLinkInventory). This makes for quick accessing of links by cell.